Fix Backmerge PR action

iangmaia opened the pull request:
## What does it do?

We have noticed the error below while creating a backmerge to a hotfix branch:

```console
Error creating backmerge pull request: git '--git-dir=<...>/.git' '--work-tree=/<...>' '-c' 'core.quotePath=true' '-c' 'color.ui=false' 'rev-parse' 'release/<HOTFIX VERSION>'  2>&1
status: pid 11499 exit 128
output: "fatal: ambiguous argument 'release/<HOTFIX VERSION>': unknown revision or path not in the working tree.\nUse '--' to separate paths from revisions, like this:\n'git <command> [<revision>...] -- [<file>...]'\nrelease/<HOTFIX VERSION>
```

Meaning that the given branch `release/<HOTFIX VERSION>` isn't present locally and only in `origin`. From inspecting the code and logs, this error most likely happened in the method `GitHelper.point_to_same_commit?`, called in the [line 106](https://github.com/wordpress-mobile/release-toolkit/blob/fc5d0311ac987bb07faf59c9ab27bece4b5b25b9/lib/fastlane/plugin/wpmreleasetoolkit/actions/common/create_release_backmerge_pull_request_action.rb#L106) of the action `create_release_backmerge_pull_request_action`.

This PRs uses `origin/` to compare the refs and some safety net to `point_to_same_commit?`.


## Checklist before requesting a review

- [x] Run `bundle exec rubocop` to test for code style violations and recommendations
- [x] Add Unit Tests (aka `specs/*_spec.rb`) if applicable
- [x] Run `bundle exec rspec` to run the whole test suite and ensure all your tests pass
- [x] Make sure you added an entry in [the `CHANGELOG.md` file](https://github.com/wordpress-mobile/release-toolkit/blob/trunk/CHANGELOG.md#trunk) to describe your changes under the appropriate existing `###` subsection of the existing `## Trunk` section.
- [x] If applicable, add an entry in [the `MIGRATION.md` file](https://github.com/wordpress-mobile/release-toolkit/blob/trunk/MIGRATION.md) to describe how the changes will affect the migration from the previous major version and what the clients will need to change and consider.

iangmaia committed:
Fetch origin prior to running the backmerge creation logic

iangmaia committed:
Add error handling to point_to_same_commit? method

dangermattic commented:
<!--
  0 Errors
  1 Warning: PR is not assigned to a milest...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iangmaia committed:
Update CHANGELOG.md

AliSoftware reviewed (approved):
Would be worth testing this in real world in a dummy repo (maybe `release-toolkit-demo`, which was created a while ago mostly for such testing purposes…?), e.g. creating a `release/12.4` branch and a `release/12.3` branch then running that action with `source_branch:'release/12.3'` and no `target_branches`)

iangmaia committed:
Always use origin for checking if the heads point to the same commit

iangmaia committed:
Revert git fetch call

iangmaia commented:
@AliSoftware 
> Would be worth testing this in real world in a dummy repo (maybe `release-toolkit-demo`, which was created a while ago mostly for such testing purposes…?), e.g. creating a `release/12.4` branch and a `release/12.3` branch then running that action with `source_branch:'release/12.3'` and no `target_branches`)

I've ran some tests, and unfortunately, fetching `origin` didn't really fix the issue, so we had to use `origin/` when comparing refs. I've updated the code on 332e9f85ae44f5b46d61c27d842f871e0b2e1a3f, and this [worked fine](https://buildkite.com/automattic/release-toolkit-demo/builds/125#0191b3fe-2008-4c02-96ad-817256b1ac91/173-239). I've also removed the origin fetch (17eb9382eb6259c6a59d7b01ff039f91961ea30b) as it seems redundant in the end.

mokagio commented on lib/fastlane/plugin/wpmreleasetoolkit/helper/git_helper.rb:
```diff
@@ -188,8 +188,14 @@ def self.fetch_all_tags
+          ref1_commit = git_repo.gcommit(ref1)
+          ref2_commit = git_repo.gcommit(ref2)
+        rescue StandardError => e
+          puts "Error: #{e.message}"
```
What do you think of:

```suggestion
          UI.error "Error fetching commits for #{ref1} and #{ref2}: #{e.message}"
```

Actually, those are two different suggestions:

- Use `UI.error` instead of `puts`
- Include the attempted operation in the error message

mokagio commented on spec/create_release_backmerge_pull_request_spec.rb:
```diff
@@ -39,7 +39,7 @@ def stub_expected_pull_requests(expected_backmerge_branches:, source_branch:, la
       next if branch_exists_on_remote
 
-      allow(Fastlane::Helper::GitHelper).to receive(:point_to_same_commit?).with(target_branch, source_branch).and_return(point_to_same_commit)
+      allow(Fastlane::Helper::GitHelper).to receive(:point_to_same_commit?).with("origin/#{target_branch}", "origin/#{source_branch}").and_return(point_to_same_commit)
```
Unrelated to this particular line, but I wonder if it would be worth adding a test to ensure that if we pass branches that already have "origin/" in the name, our logic doesn't append an additional "origin/" to them.

mokagio committed:
Merge branch 'trunk' into iangmaia/fix-backmerge-pr

mokagio commented:
@iangmaia I'll take over merging this already approved PR so that I can update the changelog etc. of my removal PR #580 on top of this. Thanks!
